### PR TITLE
Compatibility with datasets in GCS buckets

### PR DESCRIPTION
**Pull request by mmwinther**
- When a path starting with `gs://` is supplied, Datadoc will retrieve the dataset from the relevant bucket
- A `StorageAdapter` Protocol abstraction layer is written and implemented concretely for `LocalFile` and `GCSObject`
- Use of the `StorageAdapter` allows for identical implementation at the callsite regardless of underlying storage medium
- Tests with GCS mocks are written
- Authentication to GCP implemented both for Dapla and on local machine

### CODE REVIEW

File: datadoc/backend/DataDocMetadata.py [hunk 1]
```diff
@@ -38,21 +40,20 @@
     ]
 )
 
+METADATA_DOCUMENT_FILE_SUFFIX = "__DOC.json"
```
Comment (bjornandre) [inline datadoc/backend/DataDocMetadata.py:43]: Just a personal opinion, but frameworks like Spark will ignore files with a `.`-prefix (like` .metadata.json`) when reading multiple files from a folder. So maybe we should use this as well.

File: datadoc/backend/DataDocMetadata.py [hunk 2]
```diff
@@ -73,14 +74,11 @@ def __init__(self, dataset):
...
+    def get_dataset_state(self, dataset: str) -> Optional[DatasetState]:
         """Use the path to attempt to guess the state of the dataset"""
-
-        if dataset_path is None:
-            dataset_path = self.dataset_full_path
-        dataset_path_parts = list(dataset_path.parts)
+        if dataset is None:
+            return None
+        dataset_path_parts = list(pathlib.Path(dataset).parts)
         if "kildedata" in dataset_path_parts:
```
Comment (bjornandre) [inline datadoc/backend/DataDocMetadata.py:82]: This seems very fragile. Isn't it better to force the users to select the correct dataset state?

File: datadoc/backend/DataDocMetadata.py (hunk 1 above)
Comment (mmwinther) [reply datadoc/backend/DataDocMetadata.py:43]: Good suggestion, I might defer to @BjornRoarJoneid to see what he thinks about that. I personally don't have anything against it, and it also may make it more clear to users that the file should not be manually edited.

File: datadoc/backend/StorageAdapter.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    def write_text(self, text: str) -> None:
+        self._path_object.write_text(text, encoding="utf-8")
+
+    @property
+    def location(self) -> str:
+        return str(self._path_object)
+
+
+class StorageAdapter(Protocol):
```
Comment (bjornandre) [inline datadoc/backend/StorageAdapter.py:83]: Nice absraction :)

File: datadoc/backend/DataDocMetadata.py (hunk 2 above)
Comment (mmwinther) [reply datadoc/backend/DataDocMetadata.py:82]: Yes, it is fragile, I believe it is done in the interests of _trying_ to make the users' job easier (i.e. minimise how much information they must input).

On the other hand, if the path doesn't include the dataset state the only consequence is the user enters it manually.

File: datadoc/backend/StorageAdapter.py (hunk 3 above)
Comment (mmwinther) [reply datadoc/backend/StorageAdapter.py:83]: Thanks 😊

File: datadoc/backend/DataDocMetadata.py (hunk 2 above)
Comment (bjornandre) [reply datadoc/backend/DataDocMetadata.py:82]: Ok, but maybe the implementation should follow this naming standard: https://statistics-norway.atlassian.net/wiki/spaces/MPD/pages/3170140229/Navnestandard+og+versjonering+av+datasett+i+DAPLA#Navngiving-av-mapper-og-datasett-(filer)-i-DAPLAs-datalager

Which would mean that "inndata" should be more widely used than "kildedata".

File: datadoc/backend/DataDocMetadata.py [hunk 4]
```diff
@@ -79,12 +79,18 @@ def get_dataset_state(self, dataset: str) -> Optional[DatasetState]:
...
             return None
         dataset_path_parts = list(pathlib.Path(dataset).parts)
-        if "kildedata" in dataset_path_parts:
+        if "utdata" in dataset_path_parts:
+            return DatasetState.OUTPUT_DATA
+        elif "statistikk" in dataset_path_parts:
+            return DatasetState.STATISTIC
+        elif "klargjorte-data" in dataset_path_parts:
+            return DatasetState.PROCESSED_DATA
+        elif "klargjorte_data" in dataset_path_parts:
```
Comment (bjornandre) [inline datadoc/backend/DataDocMetadata.py:88]: Duplicate?

**Review by bjornandre — APPROVED** [review]

**Pull request merged**